staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -69,6 +69,251 @@ func TestSimpleQueue(t *testing.T) {
+
+	// step past the shorter item and confirm that it got queued.
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+	item, _ := q.Get()
+	q.Done(item)
+
+	// add an item into the waiting queue.
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddAfterWithOptions(second, 1*time.Second, AddAfterOptions{Existing: TakeLonger, Synchronous: true})
+	// step past the shorter item and confirm that it didn't get queued
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+
+	// Add a longer delay with the expectation of pushing the ready later...
+	q.AddAfterWithOptions(second, 20*time.Second, AddAfterOptions{Existing: TakeLonger, Synchronous: true})
+
+	// step past the original delayed item and check it queues.
+	fakeClock.Step(4 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+
+	// step past the new longer delay and prove our longer item was queued.
+	fakeClock.Step(30 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+
+	item, _ = q.Get()
+	q.Done(item)
+
+	// Show that TakeIncoming will override whatever the value of the waiting item
+	// Shorter case...
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddAfterWithOptions(second, 1*time.Second, AddAfterOptions{Existing: TakeIncoming, Synchronous: true})
+	// step past the shorter item and confirm that it didn't get queued
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+	item, _ = q.Get()
+	q.Done(item)
+
+	// Longer case...
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddAfterWithOptions(second, 20*time.Second, AddAfterOptions{Existing: TakeIncoming, Synchronous: true})
+	// step past the first waiting time...
+	fakeClock.Step(10 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+}
+
+// TestForgetDelayed - should remove an item from the delayed queue
+func TestForgetDelayed(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewDelayingQueueWithCustomClock(fakeClock, "")
+
+	first := "foo"
+
+	// add the item into the waiting queue.
+	q.AddAfterWithOptions(first, 5*time.Second, AddAfterOptions{Synchronous: true})
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
```

Comment:
why are we waiting for the timeout here? It doesn't test anything - only that we slept 5 seconds.

I see this mistake in other tests as well? can you please go over all the tests and revise them?

Reply:
In following with the existing tests, I wrote the new tests so that they check for the desired state after every action.  I agree that in this test it doesn't add much, it only ensures that the item was indeed added to the waiting queue and didn't go into the active queue by mistake.  I could change this and others but I'd like @lavalamp or one of the other active developers to confirm what they would prefer.